File: backend-comic/src/main/java/com/thjvjpxx/backend_comic/service/CategoryService.java
package com.thjvjpxx.backend_comic.service;

import java.util.List;

import com.thjvjpxx.backend_comic.dto.request.CategoryRequest;
import com.thjvjpxx.backend_comic.dto.response.BaseResponse;
import com.thjvjpxx.backend_comic.model.Category;

/**
 * Service xử lý API quản lý danh mục
 */
public interface CategoryService {
    /**
     * Lấy tất cả danh mục
     * 
     * @param page   Trang hiện tại
     * @param limit  Số lượng mỗi trang
     * @param search Tìm kiếm theo tên
     * @return Response chứa danh sách danh mục
     */
    BaseResponse<List<Category>> getAllCategories(int page, int limit, String search);

    /**
     * Lấy danh mục theo ID
     * 
     * @param id ID danh mục
     * @return Response chứa danh mục
     */
    BaseResponse<Category> getCategoryById(String id);

    /**
     * Tạo danh mục mới
     * 
     * @param category DTO chứa thông tin danh mục
     * @return Response chứa danh mục đã tạo
     */
    BaseResponse<Category> createCategory(CategoryRequest category);

    /**
     * Cập nhật danh mục
     * 
     * @param id       ID danh mục
     * @param category DTO chứa thông tin danh mục
     * @return Response chứa danh mục đã cập nhật
     */
    BaseResponse<Category> updateCategory(String id, CategoryRequest category);

    /**
     * Xóa danh mục
     * 
     * @param id ID danh mục
     * @return Response chứa danh mục đã xóa
     */
    BaseResponse<Category> deleteCategory(String id);
}

File: frontend-comic/src/services/categoryService.ts
import { CategoryCreateUpdate, CategoryResponse } from "@/types/category";
import { ApiResponse } from "@/types/api";
import { fetchApi } from "./api";

// Lấy danh sách thể loại có phân trang
export const getCategories = async (
  page: number = 1,
  limit: number = 5,
  search?: string
) => {
  let endpoint = `/categories?page=${page}&limit=${limit}`;

  if (search) {
    endpoint += `&search=${encodeURIComponent(search)}`;
  }

  return await fetchApi<CategoryResponse[]>(endpoint);
};

// Lấy chi tiết một thể loại
export const getCategory = async (
  id: number
) => {
  return await fetchApi<CategoryResponse>(`/categories/${id}`);
};

// Tạo thể loại mới
export const createCategory = async (
  data: CategoryCreateUpdate
) => {
  return await fetchApi<CategoryResponse>("/categories", {
    method: "POST",
    data: data,
  });
};

// Cập nhật thể loại
export const updateCategory = async (
  id: string,
  data: CategoryCreateUpdate
) => {
  return await fetchApi<CategoryResponse>(`/categories/${id}`, {
    method: "PUT",
    data: data,
  });
};

// Xóa thể loại
export const deleteCategory = async (
  id: string
) => {
  return await fetchApi<null>(`/categories/${id}`, {
    method: "DELETE",
  });
};

// Lấy tất cả thể loại không phân trang
export const getAllCategories = async () => {
  return await fetchApi<CategoryResponse[]>('/category');
};
